pallets/dca/src/lib.rs

```diff
@@ -0,0 +1,1107 @@
+			Error::<T>::TotalAmountShouldBeLargerThanStorageBond
+		);
+
+		Ok(())
+	}
+
+	fn ensure_schedule_is_suspended(schedule_id: ScheduleId) -> DispatchResult {
+		ensure!(
+			Suspended::<T>::contains_key(schedule_id),
+			Error::<T>::ScheduleMustBeSuspended
+		);
+
+		Ok(())
+	}
+
+	fn ensure_schedule_exists(schedule_id: &ScheduleId) -> DispatchResult {
+		ensure!(Schedules::<T>::contains_key(schedule_id), Error::<T>::ScheduleNotExist);
+
+		Ok(())
+	}
+
+	fn ensure_origin_is_schedule_owner(schedule_id: ScheduleId, who: &T::AccountId) -> DispatchResult {
+		let schedule = Schedules::<T>::get(schedule_id).ok_or(Error::<T>::ScheduleNotExist)?;
+		ensure!(*who == schedule.owner, Error::<T>::Forbidden);
+
+		Ok(())
+	}
+
+	fn ensure_sell_amount_is_bigger_than_transaction_fee(
+		schedule: &Schedule<T::AccountId, T::Asset, T::BlockNumber>,
+	) -> DispatchResult {
+		match schedule.order {
+			Order::Sell {
+				asset_in, amount_in, ..
+			} => {
+				let transaction_fee = Self::get_transaction_fee_in_sold_asset(asset_in)?;
+				ensure!(amount_in > transaction_fee, Error::<T>::TradeAmountIsLessThanFee);
+			}
+			Order::Buy { .. } => {
+				//For buy we don't check as the calculated amount in will always include the fee
+			}
+		}
+
+		Ok(())
+	}
+
+	#[transactional]
+	pub fn execute_schedule(
+		weight: &mut u64,
+		schedule_id: ScheduleId,
+		schedule: &Schedule<T::AccountId, T::Asset, T::BlockNumber>,
+	) -> DispatchResult {
+		*weight += Self::get_execute_schedule_weight();
+
+		//TODO: Add logging for each error cases
+		//TODO: TERMINATE WITH ERROR FLAG
+		let origin: OriginFor<T> = Origin::<T>::Signed(schedule.owner.clone()).into();
+
+		//If these two fail, we terminate - no rollback
+		let Ok(amount_to_unreserve)  = Self::amount_to_unreserve(&schedule.order) else {
+			return Err(Error::<T>::InvalidState.into());
+		};
+
+		//TODO: check if this returns `amount_to_unreserve`, otherwise we fail, terminate
+		let sold_currency = schedule.order.get_asset_in();
+
+		T::Currency::unreserve_named(
+			&NAMED_RESERVE_ID,
+			sold_currency.into(),
+			&schedule.owner,
+			amount_to_unreserve.into(),
+		);
+
+		let Ok(()) = Self::decrease_remaining_amount(schedule_id, amount_to_unreserve) else {
+			return Err(Error::<T>::InvalidState.into());
+		};
+
+		Self::execute_trade(origin, &schedule.order)
+	}
+
+	fn price_change_is_bigger_than_max_allowed(asset_a: T::Asset, asset_b: T::Asset) -> Result<bool, DispatchResult> {
+		let current_price = Self::get_current_price(asset_a, asset_b)?;
+		let price_from_short_oracle = Self::get_price_from_short_oracle(asset_a, asset_b)?;
+
+		let max_allowed = FixedU128::from(T::MaxPriceDifference::get());
+		let max_allowed_difference = current_price.saturating_mul(max_allowed);
+
+		let diff = if current_price > price_from_short_oracle {
+			current_price.saturating_sub(price_from_short_oracle)
+		} else {
+			price_from_short_oracle.saturating_sub(current_price)
+		};
+
+		Ok(diff > max_allowed_difference)
```

Comment:
note that this calculation is not precise enough. i rewored this in the omnipool ensure price.

use following formula:

( sp - op ) / ( (sp+op) / 2 ) <= max allowed